### PR TITLE
[quant][graphmode][fx] Support fp16 dynamic quantization for linear

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44582 [quant][graphmode][fx] Support fp16 dynamic quantization for linear**

Summary:

Test Plan:
test_quantize_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23665974](https://our.internmc.facebook.com/intern/diff/D23665974)

### CODE REVIEW

File: torch/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -538,9 +538,16 @@ def convert(self, quantizer, node, load_arg, debug=False):
...
                 # linear args:
-                # (x, weight, bias)
-                # quantize weight
-                quantized_weight = load_arg(quantized=True)(self.linear_node.args[1])
+                # (x, observed_weight, bias)
+                # get observer for the weight
+                weight_observer = quantizer.activation_post_process_map[self.linear_node.args[1].args[0].name]
+
+                if weight_observer.dtype == torch.float16:
+                    linear_weight = load_arg(quantized=False)(self.linear_node.args[1])
```
Comment (raghuramank100) [inline torch/quantization/fx/quantization_patterns.py:546]: Minor comment: It looks like the prepack_fp16 does both dtype conversion and packing, which is not true of the int8 case. Is there any way to make this uniform? i.e convert to fp16 and then call pack?

**Review by raghuramank100 — APPROVED** [review]

File: torch/quantization/fx/quantization_patterns.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:546]: right, yeah that makes sense, we can call type conversion and then prepack here if we decide to make prepack_fp16 only prepack a fp16 Tensor.

Comment (raghuramank100) [reply torch/quantization/fx/quantization_patterns.py:546]: Seems like a better engineering task to do down the line (cc @supriyar )

Comment (supriyar) [reply torch/quantization/fx/quantization_patterns.py:546]: We already have an op for type conversion. Can we add that to the graph when debug=True? `aten::_saturate_weight_to_fp16`

Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:546]: does the current prepack_fp16 do type conversion? if so we will need to remove that as well

**Pull request closed without merging**